### PR TITLE
DnD with breadcrumbs

### LINKED ISSUE #1772: File Browser Breadcrumbs as Dropzone Targets
**Is your feature request related to a problem? Please describe.**
It is currently not possible to use Drag & Drop to move a file into a parent folder. To achieve this, a user would need to use the `Move` dialog to perform the operation.

**Describe the solution you'd like**
The Breadcrumbs displayed above each file browser could be used as drop targets to allow for moving a file into a parent folder without the `Move dialog`.

https://user-images.githubusercontent.com/12774278/143065402-ca51a67c-3dc5-47e0-a8ce-3daf70f5e9af.mp4

### PR DESCRIPTION
**Pull request by tanmoyAtb**
closes #1772 

---
Submission checklist: 

#### Layout
- [x] Change looks good in the desktop web ui
- [x] Change looks good in the mobile web ui

#### Theme
- [x] Components / elements inspected in light mode 
- [x] Components / elements inspected in dark mode

### CODE REVIEW

**Review by Tbaut — CHANGES_REQUESTED** [review]
I didn't check the code yet, but testing the functionality, I got an error from the `mv` from the api.
I have a file in `/some/folder` and drag it to move it to `/some`. The request fails with
> Object { code: 400, message: "paths is required" }
The request is as follow:
```
{"paths":[],"new_path":"/some/"}
```

**Comment by tanmoyAtb** [comment]
> I didn't check the code yet, but testing the functionality, I got an error from the `mv` from the api. I have a file in `/some/folder` and drag it to move it to `/some`. The request fails with
> 
> > Object { code: 400, message: "paths is required" }
> > The request is as follow:
> 
> ```
> {"paths":[],"new_path":"/some/"}
> ```

Its working my side, I even tried with the same `/some/folder` tree. Can you check if the drag and drop on a folder in the file browser lists work instead of the breadcrumb ?

**Comment by Tbaut** [comment]
There are some weirdnesses.. I moved my file to the `/some/folder` folder, then try to DnD with breadcrumb, it fails (not on the video), I try again (visible in the video), it works, I browse to `/some`, drag the file down to `/some/folder`, it works, then try to drag it up using the breadcrumb, it fails again..


https://user-images.githubusercontent.com/33178835/148052496-5587db47-62f5-43ee-aa91-8acfa990a957.mp4

**Comment by tanmoyAtb** [comment]
> There are some weirdnesses.. I moved my file to the `/some/folder` folder, then try to DnD with breadcrumb, it fails (not on the video), I try again (visible in the video), it works, I browse to `/some`, drag the file down to `/some/folder`, it works, then try to drag it up using the breadcrumb, it fails again..
> 
>  weirdness.mp4

So I think there's some sensitivity around the way we are dragging. If you check the clip, when the move failed, there was no drag preview. Somehow the drag event failed to grab the source. 

I'll spend some time to see if I can find the actual issue.

**Comment by tanmoyAtb** [comment]
Found the issue, The useDrag wasn't re rendering when file contents changed by moving into a folder. It should be good to go now.

**Comment by asnaith** [comment]
Hey Tanmoy, I'm seeing some strange behaviour but it's kind of difficult to describe some of them. I've captured them on video so hopefully that helps. 

- I noticed there's an issue with the appearance of the breadcrumb folders after a browser refresh 

https://user-images.githubusercontent.com/3982665/148283065-80d88f18-3285-4313-a118-1bef91c62e88.mov

- When you drag a file to the folder on the breadcrumb that it's already residing in it will append a (1) to the file name and also the modal is cut short 

https://user-images.githubusercontent.com/3982665/148283304-76a58ae9-30ab-4a50-a0fe-f62c1ad76961.mov

- When you try a regular drag and drop between folders (non breadcrumb) I've noticed the move success toast elongates to the full screen (but I think it's related to these changes as I can't reproduce this on stage). 

https://user-images.githubusercontent.com/3982665/148283513-beb2114c-e4e0-45fd-b5dc-d0cb39032756.mov

- I managed to somehow get the folder name to appear in the blue element that appeared at the bottom of the screen. 

https://user-images.githubusercontent.com/3982665/148283767-5b610c72-55de-46af-85be-2dbe252eb36e.mov

**Comment by Tbaut** [comment]
There are weird CSS quirks indeed in the render built preview, but not when you build the branch locally (also there's no funky animation). I cleared the cache from render.com and redeployed. Should be live soon.

I can't reproduce the issue I had before, and from what Andrew saw, I can see the issue with moving to the folder it's already in, we should prevent this if possible

File: packages/common-components/src/Breadcrumb/Breadcrumb.tsx [hunk 1]
```diff
@@ -96,23 +114,29 @@ const Breadcrumb: React.FC<BreadcrumbProps> = ({
...
   showDropDown
 }: BreadcrumbProps) => {
   const classes = useStyles()
 
   const generateFullCrumbs = (crumbs: Crumb[]) => {
-    return crumbs.map((item: Crumb, index: number) => (
+    return crumbs.map((crumb: Crumb, index: number) => (
       <Fragment key={`crumb-${index}`}>
-        <div>
+        {!crumb.component ?  <div
```
Comment (Tbaut) [inline packages/common-components/src/Breadcrumb/Breadcrumb.tsx]: This logic is repeated 3 times in this component, and very much looks like the `FolderBreadCrumb` component to me beside some more logic. Any way we can reuse it?

Unrelated to the above comment, we don't particularly need the ternary here,  `crumb.component || <div>...</div>` would do the job, although it's maybe not as easy to read. Having the `?` at the beginning of a line is something I'm trying to enforce too :D

File: packages/common-components/src/Breadcrumb/Breadcrumb.tsx [hunk 2]
```diff
@@ -96,23 +114,29 @@ const Breadcrumb: React.FC<BreadcrumbProps> = ({
...
+          ref={crumb.forwardedRef}
+          className={clsx(crumb.active && "active", classes.wrapper)}
+        >
           <Typography
-            onClick={() => (item.onClick ? item.onClick() : null)}
-            className={clsx(classes.crumb, item.onClick && "clickable")}
+            onClick={() => (crumb.onClick ? crumb.onClick() : null)}
```
Comment (Tbaut) [inline packages/common-components/src/Breadcrumb/Breadcrumb.tsx]: No need for the ternary here either
```suggestion
            onClick={() => crumb.onClick && crumb.onClick()}
```

File: packages/common-components/src/Breadcrumb/Breadcrumb.tsx [hunk 3]
```diff
@@ -132,12 +156,17 @@ const Breadcrumb: React.FC<BreadcrumbProps> = ({
         }}
         menuItems={crumbs.map((crumb) => ({
           contents: (
-            <Typography
-              variant="body1"
-              onClick={() => (crumb.onClick ? crumb.onClick() : null)}
+            !crumb.component ?  <div
```
Comment (Tbaut) [inline packages/common-components/src/Breadcrumb/Breadcrumb.tsx]: This is one of the repetition

File: packages/common-components/src/Breadcrumb/Breadcrumb.tsx [hunk 4]
```diff
@@ -154,15 +183,18 @@ const Breadcrumb: React.FC<BreadcrumbProps> = ({
         <>
           {generateDropdownCrumb(dropdownCrumbs)}
           <div className={clsx(classes.separator)} />
-          <div>
+          {lastCrumb.component ? <div
```
Comment (Tbaut) [inline packages/common-components/src/Breadcrumb/Breadcrumb.tsx]: and this is the 3rd one.

File: packages/common-components/src/Breadcrumb/Breadcrumb.tsx [hunk 5]
```diff
@@ -171,11 +203,16 @@ const Breadcrumb: React.FC<BreadcrumbProps> = ({
...
-          onClick={() => (homeOnClick ? homeOnClick() : null)}
-        />
-        <div className={clsx(classes.separator)} />
+        <div
+          ref={homeRef}
+          className={clsx(classes.wrapper, homeActive && "active")}
+        >
+          <HomeIcon
+            className={clsx(classes.home, homeOnClick && "clickable")}
+            onClick={() => (homeOnClick ? homeOnClick() : null)}
```
Comment (Tbaut) [inline packages/common-components/src/Breadcrumb/Breadcrumb.tsx]:
```suggestion
            onClick={() => homeOnClick && homeOnClick()}
```

File: packages/files-ui/src/Utils/pathUtils.ts [hunk 6]
```diff
@@ -22,6 +22,14 @@ export function getArrayOfPaths(path: string): string[] {
   }
 }
 
+// "/" -> []
+// "/path/to/this" -> ["path", "to", "this"]
```
Comment (Tbaut) [inline packages/files-ui/src/Utils/pathUtils.ts]: If I'm not mistaking, should update the example to something like
```suggestion
// [] -> "/"
// ["path", "to", "this"] => "/path/to/this"
```

**Review by Tbaut — COMMENTED** [review]
I won't lie, the useDrop is pretty magical to me still as I've never played with it. My comments on the code are therefore on everything else :D

File: packages/common-components/src/Breadcrumb/Breadcrumb.tsx (hunk 3 above)
Comment (tanmoyAtb) [reply packages/common-components/src/Breadcrumb/Breadcrumb.tsx]: Sorry about these, typing habits !
We'll need to find a way of formatting these with eslint.

**Comment by tanmoyAtb** [comment]
Alright I've updated The current folder in breadcrumb to not accept drops!

But the breadcrumbs breaking up is a serious issue. Look at this one. This happens when you build it even locally.

Screenshot 2022-01-07 at 12 14 30 AM
!

What happens is `JSS`  re initialises classnames to the dynamic component I'm passing as `component` to a `crumb`. 
The passed in breadcrumbs have `jss1` and `jss2` which are existing classnames which are injecting their styles !!!!!

I found this thread on the `jss` repo.
https://github.com/cssinjs/jss/issues/1458

A couple of hacky solutions in my mind i'll need to check.

**Comment by Tbaut** [comment]
Wow this is wild. One thing that crossed my mind and you prob thought about, what if we prevent the classes from having generated names, and actually go with our own naming like: `className={clsx(classes.duplicated, "unique-identifier")}` and then
```
{
  duplicated: {
    "&.unique-identifier" : {
      ...
    }
  }
}
```

**Comment by tanmoyAtb** [comment]
> Wow this is wild. One thing that crossed my mind and you prob thought about, what if we prevent the classes from having generated names, and actually go with our own naming like: `className={clsx(classes.duplicated, "unique-identifier")}` and then
> 
> ```
> {
>   duplicated: {
>     "&.unique-identifier" : {
>       ...
>     }
>   }
> }
> ```

This would have worked, Yes !
But I found the solution in the material docs:
https://mui.com/styles/api/
https://github.com/mui-org/material-ui/issues/11843

The PR seems to be in order now !

File: packages/files-ui/src/Components/Modules/FileBrowsers/views/FilesList.tsx [hunk 7]
```diff
@@ -800,8 +830,24 @@ const FilesList = ({ isShared = false }: Props) => {
...
+                handleMove={(item) => {
+                  moveItems && crumb.path && moveItems(item.ids, crumb.path)
+                  setSelectedItems([])
+                }}
+                handleUpload={(item) => handleUploadOnDrop &&
+                  crumb.path &&
+                  handleUploadOnDrop(item.files, item.items, crumb.path)}
+              /> : null
+            }))
+            }
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/FileBrowsers/views/FilesList.tsx]: Just a nit to have the ? and : at the beginning of the line, and closing all the } and ) on 1 line.
```suggestion
              component: (i < crumbs.length - 1)
                ? <FolderBreadcrumb
                  folderName={crumb.text}
                  onClick={crumb.onClick}
                  handleMove={(item) => {
                    moveItems && crumb.path && moveItems(item.ids, crumb.path)
                    setSelectedItems([])
                  }}
                  handleUpload={(item) => handleUploadOnDrop &&
                  crumb.path &&
                  handleUploadOnDrop(item.files, item.items, crumb.path)}
                />
                : null
            }))}
```

**Review by Tbaut — APPROVED** [review]
Played again with it, including moving and uploading files, and it worked like a charm :tada: 
The code looks neat too, just left a small comment for a nit.

**Comment by asnaith** [comment]
Awesome work Tanmoy, I can't see any of the previous weirdness and the feature is working great for me too

**Review by asnaith — APPROVED** [review]

**Pull request merged**